python/device/measurement.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for
+# full license information.
+import statistics
+import threading
+import contextlib
+import datetime
+
+
+class ThreadSafeCounter(object):
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.value = 0
+
+    def set(self, value):
+        with self.lock:
+            self.value = value
+
+    def add(self, value):
+        with self.lock:
+            self.value += value
+
+    def increment(self):
+        self.add(1)
+
+    def decrement(self):
+        self.add(-1)
+
+    def get_count(self):
+        with self.lock:
+            return self.value
+
+    def extract_count(self):
+        with self.lock:
+            value = self.value
+            self.value = 0
+            return value
+
+
+class ThreadSafeList(object):
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.list = []
+
+    def append(self, value):
+        with self.lock:
+            self.list.append(value)
+
+    def clear(self):
+        with self.lock:
+            self.list.clear()
+
+    def get_average(self):
+        with self.lock:
+            return statistics.mean(self.list)
+
+    def extract_average(self):
+        with self.lock:
+            average = 0
+            if len(self.list):
+                average = statistics.mean(self.list)
+                self.list.clear()
+            return average
+
+    def extract_list(self):
+        with self.lock:
+            old_list = self.list
+            self.list = []
+            return old_list
+
+
+class MeasureLatency(contextlib.AbstractContextManager):
+    def __init__(self, tracker=None):
+        self.start_time = None
+        self.end_time = None
+        self.tracker = tracker
+
+    def __enter__(self):
+        self.start_time = datetime.datetime.now()
```

Comment:
utc?

Reply:
Good point  We should always use UTC when pushing data up.  This class is about measuring latency.  I may even change this to use epoch time since we only care about deltas in seconds.  

---
In reply to: [474140237](https://github.com/Azure/iot-sdk-longhaul/pull/3#discussion_r474140237) [](ancestors = 474140237)